monkey/common/configuration/agent_configuration.py

```diff
@@ -0,0 +1,194 @@
+    timeout = fields.Float()
+
+    @post_load
+    def _make_icmp_scan_configuration(self, data, **kwargs):
+        return ICMPScanConfiguration(**data)
+
+
+@dataclass(frozen=True)
+class TCPScanConfiguration:
+    timeout: float
+    ports: List[int]
+
+
+class TCPScanConfigurationSchema(Schema):
+    timeout = fields.Float()
+    ports = fields.List(fields.Int())
+
+    @post_load
+    def _make_tcp_scan_configuration(self, data, **kwargs):
+        return TCPScanConfiguration(**data)
+
+
+@dataclass(frozen=True)
+class NetworkScanConfiguration:
+    tcp: TCPScanConfiguration
+    icmp: ICMPScanConfiguration
+    fingerprinters: List[PluginConfiguration]
+    targets: ScanTargetConfiguration
+
+
+class NetworkScanConfigurationSchema(Schema):
+    tcp = fields.Nested(TCPScanConfigurationSchema)
+    icmp = fields.Nested(ICMPScanConfigurationSchema)
+    fingerprinters = fields.List(fields.Nested(PluginConfigurationSchema))
+    targets = fields.Nested(ScanTargetConfigurationSchema)
+
+    @post_load
+    def _make_network_scan_configuration(self, data, **kwargs):
+        return NetworkScanConfiguration(**data)
+
+
+class ExploitationOptionsConfigurationSchema(Schema):
+    http_ports = fields.List(fields.Int())
+
+    @post_load
+    def _make_exploitation_options_configuration(self, data, **kwargs):
+        return ExploitationOptionsConfiguration(**data)
+
+
+@dataclass(frozen=True)
+class ExploiterConfiguration:
+    name: str
+    options: Dict
+    supported_os: List[OperatingSystems]
+
+
+class ExploiterConfigurationSchema(Schema):
+    name = fields.Str()
+    options = fields.Mapping()
+    supported_os = fields.List(EnumField(OperatingSystems))
+
+    @post_load
+    def _make_exploiter_configuration(self, data, **kwargs):
+        return ExploiterConfiguration(**data)
+
+
+@dataclass(frozen=True)
+class ExploitationConfiguration:
+    options: ExploitationOptionsConfiguration
+    brute_force: List[ExploiterConfiguration]
+    vulnerability: List[ExploiterConfiguration]
+
+
+class ExploitationConfigurationSchema(Schema):
+    options = fields.Nested(ExploitationOptionsConfigurationSchema)
+    brute_force = fields.List(fields.Nested(ExploiterConfigurationSchema))
+    vulnerability = fields.List(fields.Nested(ExploiterConfigurationSchema))
+
+    @post_load
+    def _make_exploitation_options_configuration(self, data, **kwargs):
+        return ExploitationConfiguration(**data)
+
+
+@dataclass(frozen=True)
+class PropagationConfiguration:
+    maximum_depth: int
+    network_scan: NetworkScanConfiguration
+    exploitation: ExploitationConfiguration
+
+
+class PropagationConfigurationSchema(Schema):
+    maximum_depth = fields.Int()
+    network_scan = fields.Nested(NetworkScanConfigurationSchema)
+    exploitation = fields.Nested(ExploitationConfigurationSchema)
+
+    @post_load
+    def _make_propagation_configuration(self, data, **kwargs):
+        return PropagationConfiguration(**data)
```

Comment:
This boilerplate can probaby extend `Schema`. Not 100% sure it's worth the trouble though

Reply:
I'm not sure there's really a good solution. We'll have to build something confusing to save 3 lines.